### DISCUSSION TITLE
Scenario Selective Inheritance (SSI) - Technical Proposal Draft.

### DISCUSSION THREAD

**Discussion by francescoloconte** [FIXTradingCommunity/fix-orchestra#199] [Ideas]
# Scenarios and Inheritance

Scenarios in Orchestra allow for defining variations of entities. With Orchestra 1.0, a scenario is another instance of its base entity, only sharing their id and name, without other entity relationship. To improve the value of scenarios, a scenario relationship is proposed. This enables a scenario to inherit the structure and attributes of another, modify the structure by adding, removing, and replacing elements, and adjust attributes while retaining a parent/child relationship, where changes to the parent entity are reflected in the child scenarios, similarly to inheritance in object-oriented design.

To implement this inheritance, the addition of `inheritsScenarioId` attribute to the `oidGrp` is recommended, applicable to all container types, as well as `fieldType` and `codeType`.

## Scenario Selective Inheritance

In the context of object-oriented design and development, inheritance is a mechanism where a new class (subclass) is created, absorbing, and extending the characteristics and behaviors (attributes and methods) of an existing class (superclass). This allows the subclass to reuse code from the superclass and add new functionalities or modify existing ones as needed.

With this type of inheritance is not possible to remove or hide what is being inherited from the superclass. However, in the case of Orchestra, is it desirable to *selectively* inherit elements from the parent scenario, or in other words, remove inherited elements that are not desired. For this reason, we call our technique Scenario Selective Inheritance (or SSI) since it allows selecting what to inherit from the parent scenario.

*[It is important to clarify that the technique does not provide pure inheritance because, by definition, inheritance would require that a child scenario be usable anywhere the parent scenario is usable. Since we allow removal of elements in the child scenario, this is no longer possible. Henceforth the need to use an adequate name.]*

## List of Terms

- *Parent Scenario, Super Scenario: A scenario to inherit from. A Child/Sub Scenario inherits from it.*
- *Child Scenario, Sub Scenario: A scenario inheriting from a Parent/Super Scenario.*

## Element Additions

A Child Scenario can add new elements that are not present in its Parent Scenario. This is done by referencing the new elements in the scenario definition.

**Example**

Define Scenario 2 for the Instrument component, adding the MaturityDate field:

```
<fixr:component id="1003" name="Instrument" scenarioId="1">
    <fixr:fieldRef id="55" name="Symbol"/>
    <fixr:fieldRef id="48" name="SecurityID/"/>
    <fixr:fieldRef id="22" name="SecurityIDSource"/>
    <fixr:fieldRef id="107" name="SecurityDesc"/>
    <fixr:fieldRef id="200" name="MaturityMonthYear"/>
</fixr:component>

<fixr:component id="1003" name="Instrument" scenarioId="2" inheritsScenarioId=”1”>
    <fixr:fieldRef id="541" name="MaturityDate"/>
</fixr:component>
```

### Positional Element Addition

Certain encodings, such as FIX *tag=value* encoding, rely on the ordering of elements, for example, the order of fields in the message. When creating a scenario, it is therefore important to be able to specify where new elements should be added. To address this requirement, we propose a simple approach that uses other elements as *anchor points*.

The method for placing a new element involves referencing the adjacent element in the parent scenario (*anchor*), indicating where to position the new element (*pivot*) relative to the *anchor* (e.g., before, after, etc.). To achieve this, the following attributes are suggested for referencing existing elements of `messageType`, `componentType`, `groupType`, or `codeSetType`:

- `position`: where the new element should be placed, e.g., before the referenced element
- `positionRefType`: the type of the referenced element (component, group, or field)
- `positionRefId`: the identifier of the referenced element
- `positionRefScenarioId`: the optional identifier of the scenario of the referenced element.

**Example**

Add the tag 20002 after the tag 44:

```
<fieldRef id="20002" name="Client2UDF" position="after" positionRefType="field" positionRefId="44"/>
```

Since it is not uncommon to add new elements as the first or last one in a container, the position attribute enumeration should include two more options: `first` and `last`.

When no position attribute is specified, the element is added as the last one in the container.

#### Note on `instanceName` and `positionRefInstanceName`

In Orchestra, when defining containers such as messages, components, or groups, their constituent elements can reference the same element more than once. For example, a message can reference a field twice (recurring elements). Since references do not have unique identifiers, it would not be possible to tell them apart. Therefore, Orchestra provides the `instanceName` attribute on field reference and component references. This provides further instances of references (to the same element) with a locally unique name.

It follows that if a Parent Scenario includes recurring elements, it is necessary to distinguish between them to use them as anchor points. To address this, an additional optional attribute is proposed:

- `positionRefInstanceName`: the optional instance name of the referenced element

The additional attribute should match the `instanceName` of the referenced element. If there are recurring elements and `positionRefInstanceName` is not specified, it defaults to the original one element.

**Effect on CodeSet Sort Order**

The Code elements in a CodeSet have an attribute called *sort* to sort them for displaying purposes. However, by supporting scenarios for a CodeSet, when a child scenario adds new Codes, it will not be able to use the *sort* attribute to interleave them with existing Codes, since sort numbers are contiguous. This renders the *sort* attribute ineffective.

We recommend therefore deprecating the *sort* attribute from the Code element, and instead making their order of definition meaningful, even for displaying purposes, just like for fields in messages. It will be possible to use the technique described above to interleave Codes.

**Example**

*Note: nested element* `structure` *of* `message` *is omitted for simplicity.*

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
    <fixr:fieldRef id="54" name="Side"/>
    <fixr:fieldRef id="40" name="OrdType"/>
    <fixr:fieldRef id="40" name="OrdType" instanceName="OrdType2"/>
    <fixr:fieldRef id="38" name="OrderQty"/>
    <fixr:fieldRef id="44" name="Price"/>
    <fixr:componentRef id="1012" name="Parties"/>
    <fixr:fieldRef id="1" name="Account" scenarioId="2"/>
</fixr:message>

<fixr:message msgType="D" id="14" name="NewOrderSingle" scenarioId="2" inheritsScenarioId="1">

    <!-- After the specified field -->
    <fixr:fieldRef id="20002" name="Client2UDF" position="after" positionRefType="field" positionRefId="44"/>

    <!-- After the specified field, refining the name -->
    <fixr:fieldRef id="20003" name="Client2KLC" position="after" positionRefType="field" positionRefId="40" positionRefInstanceName="OrdType2"/>

    <!-- Before the specified field, which references the scenario -->
    <fixr:fieldRef id="20013" name="Client2NAC" position="before" positionRefType="field" positionRefId="1" positionRefScenarioId="2"/>

    <!-- Before the specified component -->
    <fixr:fieldRef id="20009" name="Client2IDK" position="before" positionRefType="component" positionRefId="1012"/>

    <!-- After Price in this order -->
    <fixr:fieldRef id="20027" name="NewField1" position="after" positionRefType="field" positionRefId="44"/>
    <fixr:fieldRef id="20028" name="NewField2" position="after" positionRefType="field" positionRefId="44"/>
    <fixr:fieldRef id="20029" name="NewField3" position="after" positionRefType="field" positionRefId="44"/>

    <!-- At the very beginning of the message (container) -->
    <fixr:fieldRef id="20010" name="Client2ADK" position="first"/>

    <!-- At the very end of the message (container) -->
    <fixr:fieldRef id="20011" name="Client2DKE" position="last"/>

    <!—This is the same as the one above, adding the field to the end of the message
        <fixr:fieldRef id="20011" name="Client2DKE"/>
     -->
</fixr:message>
```

#### Clarifications on Usage

Consider the following scenario with fields labeled F1, F2, F3, F4, and F5:

**Adding After:**

We start with this structure:

- F1 → F5

Adding F2 after F1, using F1 as the anchor point, results in this:

- F1 → F2 → F5

Adding F3 after F1 should result in:

- F1 → F2 → F3 → F5

This method ensures that new fields are incorporated in the order they appear in XML.

**Adding Before:**

Staring again with this structure:

- F1 → F5

Adding F2 before F5, using F5 as the anchor point, results in this:

- F1 → F2 → F5

Then, including F3 before F5 should result in this:

- F1 → F2 → F3 → F5

Note that in the final result, fields appear in the same descending order as they were added.

**Interleaving:**

Let's consider the scenario of adding F2 after F1 and F3 before F5:

- F1 → F2 → F3 → F5

Lastly, adding F2 after F1, F3 before F5, and F4 after F1 results in:

- F1 → F2 → F4 → F3 → F5

Following the algorithm, F4 is added after F2, avoiding any interleaving.

**Schema Changes:**

In `fieldRefType`, `componentRefType`, `groupRefType`, `codeType` add the following optional attributes:

```
• position="after|before|first|last"
• positionRefType="field|component|group" (except for codeType)
• positionRefId="<fixr:id\_t>"
• positionRefScenarioId="<fixr:id\_t>"
• positionRefInstanceName="<fixr:Name\_t>"
```

In `codeType`, remove the `sort` attribute.

- **Advantages**: ease of maintenance, flexibility, readability
- **Disadvantages**: verbosity

## Element Removals

A Sub Scenario can remove elements that are present in its Parent Scenario. Similarly to how `fieldRef` is used to add (by reference) a field, and code is used to define a new value in a codeSet, we propose introducing new element types `fieldUnref`/`groupUnref`/`componentUnref` and `codeUndef` to *unreference* and *undefine* elements and code values respectively.

To identify the element to be removed, the new XML elements would have the following attributes:

- `id`: The `id` of the element being removed.
- `scenarioId`: The `scenarioId` of the element being removed.
- `instanceName`: [Optional] The `instanceName` of the element being removed. Required to differentiate the same element appearing multiple times in the parent scenario.

**Example**

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
    <fixr:fieldRef id="11" name="ClOrdID"/>
    <fixr:fieldRef id="55" name="Symbol"/>
    <fixr:fieldRef id="54" name="Side"/>
    <fixr:fieldRef id="40" name="OrdType"/>
    <fixr:fieldRef id="40" name="OrdType" instanceName="OrdType2"/>
    <fixr:fieldRef id="38" name="OrderQty"/>
    <fixr:fieldRef id="44" name="Price"/>
    <fixr:componentRef id="1012" name="Parties"/>
    <fixr:fieldRef id="1" name="Account"/>
</fixr:message>

<fixr:message msgType="D" id="14" scenarioId="2" inheritsScenarioId="1">
	<fixr:fieldUnref id="40" name="OrdType2"/>
	<fixr:componentUnref id="1012"/>
</fixr:message>
```

**Example**

```
<fixr:codeSet type="char" id="13" name="CommTypeCodeSet">
	<fixr:code value="1" sort="1" id="13001" name="PerUnit"/>
	<fixr:code value="2" sort="2" id="13002" name="Percent"/>
	<fixr:code value="3" sort="3" id="13003" name="Absolute"/>
	<fixr:code value="4" sort="4" id="13004" name="BasisPoints"/>
	<fixr:code value="5" sort="5" id="13005" name="AmountPerContract"/>
</fixr:codeSet>

<fixr:codeSet type="char" id="13" scenarioId="2" inheritsScenarioId="1">
	<fixr:codeUndef id="13004" name="BasisPoints"/>
</fixr:codeSet>
```

#### Schema Changes

**Schema Changes:**

In `messageType`/`structure`, `componentType`, `groupType` add `fieldUnref`, `groupUnref`, `componentUnref` elements.

In `codeSetType` add `codeUndef` element.

Add the following attributes to the new elements:

```
• id="<fixr:id\_t>"
• scenarioId="<fixr:id\_t>"
• instanceName="<fixr:Name\_t>"
```

### Notes on `supported` and `presence` attributes

While removal of elements could be accomplished by manipulating and reusing the existing `supported` or `presence` attributes, we recommend avoiding using these attributes, since they already have a clear semantic, and instead introducing the dedicated approach described above.

## Element Replacements

A Child Scenario can replace elements that are present in its Parent Scenario. A straightforward approach involves first removing the element being replaced, and subsequently adding it again with the desired modifications. The replaced element can be inserted at the old position or to a new position. If none is specified, it is added as the last entry in the scenario.

### Explicit Replacement

Elements are first removed and then added again.

**Example**:

Redefine a code by altering its value and name.

```
<fixr:codeSet type="char" id="13" name="CommTypeCodeSet" scenarioId="2" inheritsScenarioId="1">
	<fixr:codeUndef id="13006" name="PointsPerBondOrContract"/>
	<fixr:code id="13006" name="PointsPerBondOrContract_New" value="60" sort="6" position="after" positionRefId="13005"/>
</fixr:codeSet>
```

**Example**:

Replace field 44 with field 66.

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
	<fixr:fieldRef id="11" name="ClOrdID"/>
	<fixr:fieldRef id="55" name="Symbol"/>
	<fixr:fieldRef id="54" name="Side"/>
	<fixr:fieldRef id="40" name="OrdType"/>
	<fixr:fieldRef id="38" name="OrderQty"/>
	<fixr:fieldRef id="44" scenarioId=”1” name="Price"/>
	<fixr:fieldRef id="1" name="Account"/>
</fixr:message>

<fixr:message msgType="D" id="14" name="NewOrderSingle" scenarioId="2" inheritsScenarioId="1">
	<fixr:fieldUnref id="44"/>
	<fixr:fieldRef id="66" position=”after” positionRefType=”field” positionRefId=”38”/>
</fixr:message>
```

### Implicit Replacement

Since `scenarioId` is part of the element's unique key, altering it would essentially replace the element. However, changing the scenario of an element within a child scenario of its container is a common task, and the implicit replacement technique is tailored for this purpose.

If the element needing a `scenarioId` change can be referenced solely by an `id` without ambiguity, the new `scenarioId` can be specified along with other modifications. In such cases, there's no requirement to remove the original element beforehand.

**Example**:

Field 44 has scenarios 1 and 2, and we want to change from 1 to 2, and update its presence too.

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
	<fixr:fieldRef id="11" name="ClOrdID"/>
	<fixr:fieldRef id="55" name="Symbol"/>
	<fixr:fieldRef id="54" name="Side"/>
	<fixr:fieldRef id="40" name="OrdType"/>
	<fixr:fieldRef id="38" name="OrderQty"/>
	<fixr:fieldRef id="44" scenarioId=”1” name="Price"/>
	<fixr:fieldRef id="1" name="Account"/>
</fixr:message>

<fixr:message msgType="D" id="14" name="NewOrderSingle" scenarioId="2" inheritsScenarioId="1">
	<fixr:fieldRef id="44" scenarioId=”2” presence="required"/>
</fixr:message>
```

## Element Attribute Changes

When a child scenario of an entity is defined, its attributes are inherited from the parent scenario. Any attribute except ones from `oidGrp` can be overridden in the child scenario. However, in most cases, some attributes, like `msgType` or `category`, should not be changed.

If the child scenario of a container (e.g., `component`) includes a reference which matches an existing reference in the parent scenario by `id`, `scenarioId`, and `instanceName`, then all the rest of the attributes are inherited from that reference. Any attribute except `id` and `name` can be overridden. Changing the scenario of the reference is a special case of element replacement addressed in section 6.2.

Annotations, rules, and assignments are inherited in the same way, and can be overridden by redefining them in a child scenario.

**Example**

Change optional field to required, change scenario and synopsis of component.

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
	<fixr:fieldRef id="11" name="ClOrdID"/>
	<fixr:fieldRef id="55" name="Symbol"/>
	<fixr:fieldRef id="54" name="Side"/>
	<fixr:fieldRef id="40" name="OrdType"/>
	<fixr:fieldRef id="38" name="OrderQty"/>
	<fixr:fieldRef id="44" name="Price"/>
	<fixr:componentRef id="1012" name="Parties">
	<fixr:annotation>
		<fixr:documentation purpose="SYNOPSIS">Parties component description</fixr:documentation>
	</fixr:annotation>
	</fixr:componentRef>
	<fixr:fieldRef id="1" name="Account"/>
</fixr:message>

<fixr:message msgType="D" id="14" name="NewOrderSingle" scenarioId="2" inheritsScenarioId="1">
	<fixr:fieldRef id="44" name="Price" presence="required"/>
	<fixr:componentRef id="1012" scenarioId="2">
		<fixr:annotation>
			<fixr:documentation purpose="SYNOPSIS">Specific Parties component description</fixr:documentation>
		</fixr:annotation>
	</fixr:componentRef>
</fixr:message>
```

It's important to note that changing the `scenarioId`  entails nuances since it involves replacing the element reference with another one (refer to the "Implicit Replacement" section above).

**Example**

Define a scenario of a field with a constraint to its length, and a modified annotation, inheriting all other attributes.

```
<fixr:field name="Symbol" type="String" id="55">
	<fixr:annotation>
		<fixr:documentation purpose="SYNOPSIS">
			Ticker symbol.
		</fixr:documentation>
	</fixr:annotation>
</fixr:field>

<fixr:field name="Symbol" id="55" implLength="4" scenarioId="2" inheritsScenarioId="1">
	<fixr:annotation>
		<fixr:documentation purpose="SYNOPSIS">
			Four-character ticker symbol.
		</fixr:documentation>
	</fixr:annotation>
</fixr:field>
```

## Handling Parents Changes

When changes are made to a Parent Scenario, its Children Scenarios will be affected as it follows:

- If an element is added to the Parent Scenario, the new element will appear in Child Scenarios.
- If an element is removed from the Parent Scenario, the element will disappear from Child Scenarios. In this case, all references to this element present in the Children Scenarios must be removed. This is only necessary where the removed element is used as an anchor point by other references. These may need to be adjusted after the removal.
- If an element is modified in the Parent Scenario, the modifications will appear in its Child Scenarios.
- If a Parent Scenario is deleted, Child Scenarios can either inherit from another scenario, with the necessary adjustments, or from none.

## Algorithm for SSI Processing

When processing the definition of a scenario, the following rules should be observed.

1. **First Pass**:
   1. Starting at the top, read the scenario definition. As references are encountered, they are added to an *After* and *Before* list maintained for each anchor point. Lists are ordered (top-down). Separate *Top*/*Bottom* lists are maintained for elements added at the top and bottom of the scenario.
   2. **Adding After**: When field A is added after field X, it is put in the *After* list maintained for field X. When field B is also added after field X, it is added to the same list.
   3. **Adding Before**: When field C is added before field X, it is put in the *Before* list maintained for field X. When field D is also added before field X, it is added to the same list.
   4. **Adding First**: When field E is added to the top of the scenario, it is put in the *Top* list maintained for the scenario, and so forth if more are added.
   5. **Adding Last**: When field F is added to the bottom of the scenario, field F is put in the *Bottom* list maintained for the scenario, and so forth if more are added.
2. **Second Pass**:
   1. Once instructions to add elements have been processed, elements in each list are inserted at the specified position. Elements in the *After* list are added after the specified anchor point and elements in the *Before* list are added before the specified anchor point. Elements from the *Top* list are added to the very top of the scenario and elements in the *Bottom* list are added to the very bottom. References to remove or modify elements are then processed. Elements added from lists will read top down in the same order as they were specified in the scenario.
3. **Other Rules**:
   1. Elements being added cannot be used as anchor points, as they do not exist in the scenario structure until final processing.
   2. Elements being removed cannot be used as anchor points in the scenario.

## Comparison With Previous Proposals

A previous technical proposal called "Include Scenarios with Exceptions”, in section 4.3 allowed using scenarios as building blocks to other scenarios, similarly to how components are used to build messages. This would enable for example the use of different scenarios of the same element (e.g., A and B) to build a new scenario of the element (e.g., C = A + B), while providing control over the placement of new elements, but also requiring removal of duplicate elements from the addends.

While this technique offers flexibility, it approaches scenarios as building blocks for other scenarios, similarly to how existing FIX components are used. To achieve this, a scenario would not inherit anything from its parent, but instead it would re-define its structure using data from the parent, through direct references to it (as a building block).

In our opinion, this technique approaches scenarios differently from their original ethos, where scenarios are assumed to be slightly different versions of their parent.

We think the techniques described in this paper (Scenarios Selective Inheritance, or SSI), where scenarios inherit the structure of their parent, which is then modified with simple instructions, is simpler.

A criticism voiced toward SSI is that it could be used to modify scenarios to such an extent they have no resemblance to their parent. However, the potential for misuse of any technique is impossible to remove, and in fact this is already possible with the current Orchestra scenarios. However, since SSI instantiates scenarios with the full structure inherited from their parent, it has the benefit of precluding scenarios from being used as a building blocks.

Additionally, when adding fields to the middle of a scenario, the previous proposal required breaking the parent scenario into parts first. Our proposal offers a simpler approach.

Additionally, SSI has the following additional advantages:

1. Allows altering element’s attributes, such as changing a field’s presence, when defining an element scenario.
1. Introduces the ability to control the ordering of elements easily.
1. Introduces support for Field’s scenarios inheritance.
1. Introduces support for changing the scenario of a Field when referencing it from a scenario.

**Comparison Table**

|**Inclusion With Exceptions**|**Restriction and Extension**|
| --- | --- |
| **Schema changes**<ul><li>Add element `scenarioRef` to the possible elements of `messageType`, `componentType`, `groupType`, `codeSetType`.</li><li>Add element `exclude` inside of the element `scenarioRef` to restrict the elements to be included from the scenario.</li></ul> | **Schema changes**<ul><li>Add attribute `baseScenario` to `messageType`, `componentType`, `groupType`, `codeSetType`.</li><li>Add elements `fieldUnref`, `componentUnref`, `groupUnref` to the possible elements of `messageType`/`structure`, `componentType`, `groupType`.</li><li>Add element `codeUndef` to the possible elements of `codeSetType`.</li><li>Add attributes `position`, `positionRefId`, `positionRefScenarioId`, `positionRefName` to `fieldRefType`, `componentRefType`, `groupRefType`, `codeType`.</li><li>Add attribute `positionRefType` to `fieldRefType`, `componentRefType`, `groupRefType`.</li></ul> |
| **Advantages**<ul><li>Ability to include multiple scenarios.</li><li>Fewer schema changes.</li></ul> | **Advantages**<ul><li>Clean parent-child relationship in scenarios.</li><li>Precise control of field ordering in scenarios.</li></ul> |
| **Disadvantages**<ul><li>Complex structure if used with exclusions.</li><li>Composition instead of inheritance, which makes scenarios look like components.</li></ul> | **Disadvantages**<ul><li>More schema changes: several new attributes related to scenario inheritance.</li></ul> |

**Important Note**

Both proposals support the design of scenario using solely restrictions or extensions. Both approaches also require adjustments to child scenarios when parent scenarios are modified (or referenced scenarios, in the case of the previous approach).

**Comparison Example (MaturityDate only in scenario B but not Symbol)**

Shows how the same result can be achieved with both approaches.

**Parent Scenario:**

```
<fixr:component id="1003" name="Instrument" scenarioId="1">
    <fixr:fieldRef id="55" name="Symbol"/>
    <fixr:fieldRef id="48" name="SecurityID"/>
    <fixr:fieldRef id="22" name="SecurityIDSource"/>
    <fixr:fieldRef id="107" name="SecurityDesc"/>
    <fixr:fieldRef id="200" name="MaturityMonthYear"/>
</fixr:component>
```

**Child Scenario (Previous Proposal):**

```
<fixr:component id="1003" name="Instrument" scenarioId="2">
    <fixr:scenarioRef name="A">
        <fixr:exclude id="55" name="Symbol"/>
    </fixr:scenarioRef>
    <fixr:fieldRef id="541" name="MaturityDate"/>
</fixr:component>
```

**Child Scenario (Current Proposal):**

```
<fixr:component id="1003" name="Instrument" scenarioId="2" inheritsScenarioId="1">
    <fixr:fieldUnref type="field" id="55" name="Symbol"/>
    <fixr:fieldRef id="541" name="MaturityDate"/>
</fixr:component>
```

### Multiple Reference Instances

In our opinion, the `instanceName` attribute may not be necessary and we recommend reconsidering its design, probably in favor of using the `name` attribute for the task.

The `instanceName` attribute is used, for example, when a field is referenced multiple times in a component, and it is desirable to have a way to differentiate among these references. For example, to allow a scenario to use these recurring fields as anchor points.

We recommend that this attribute be deprecated, and instead one overriding the existing `name` attribute for recurring fields be used in its place. This should be possible because the name attribute is not actually part of an element key, which is composed of `element type` + `element id` + `element scenario id`.

### Notes on Elements Sorting

*Here we explain another possible approach that we do not recommend. This section can be skipped.*

A similar solution to address the sorting of element could have been to add a new sorting attribute to the `fieldRefType`, `componentRefType`, and `groupRefType` entities. Similarly to the existing `sort` integer property in `codeType`, we could introduce an attribute to these types to sort newly added elements. When adding new elements to a Child Scenario, this attribute would indicate where the new element is placed in the Parent Scenario by indicating a sort order.

**Example**

*In the example below we show how elements in the parent scenario use widely spaced sort numbers to allow the insertion of new elements by child scenario.*

```
<fixr:message msgType="D" id="14" name="NewOrderSingle">
    <fixr:fieldRef id="40" name="OrdType" sort="40"/>
    <fixr:fieldRef id="38" name="OrderQty" sort="50"/>
	<fixr:fieldRef id="44" name="Price" sort="60"/>
	<fixr:componentRef id="1012" name="Parties" sort="70"/>
	<fixr:fieldRef id="1" name="Account" sort="80"/>
</fixr:message>

<fixr:message msgType="D" id="14" name="NewOrderSingle" scenarioId="2" inheritsScenarioId="1">
    <!-- After tag 44, the preferred way -->
    <fixr:fieldRef id="20002" name="Client2UDF" sort="65"/>

    <!-- After tag 44, but leaves no space between them -->
    <fixr:fieldRef id="20003" name="Client2FDU" sort="60"/>   	

    <!-- After last tag of base scenario -->
    <fixr:fieldRef id="20011" name="Client2DKE"/>

    <!-- At the very beginning of the message (container) -->
    <fixr:fieldRef id="20010" name="Client2IDK" sort="0"/>

    <!-- At the very end of the message (container) -->
    <fixr:fieldRef id="20015" name="Client2HBN" sort="999"/>
</fixr:message>
```

If absent, the element’s `sort` value would be set to the last element in the parent scenario. Using the same `sort` value of an existing element in the parent scenario would result in the new element appearing immediately after it, potentially restricting placement options in subsequent scenario layers.

However, we feel that this approach is rigid because it requires planning of sort numbers in the parent scenario, and therefore we do not recommend it.

**Comment by JoanPuig** [comment]
@francescoloconte: Thanks for working on this, the proposal looks great!

I believe it covers all requirements we have been discussing in the calls. It also is much simpler than alternative proposals. 

The section explaining the algorithm on how scenarios must be constructed should ensure users and vendors are never surprised with the resulting outcome of an inherited object.

I like the name because as mentioned in the document it is not pure inheritance, so developers coming from other technologies will fin the emphasis on that distinction useful at the beginning. Also might be worth mentioning, that if someone is not constrained by needing a specific order (FIX Tag Value encoding), they can get pure inheritance at the Message and Component level by just using additions. 

Agreed that sort needs to be deprecated in CodeSets. I would remove the alternative sorting mechanism from this doc for brevity and clarity. Instead, I suggest to move it to the Comparison with Previous AND Alternatives Proposals at the end.

Small typo: After the interleaving explanation there are "Schema Changes" twice.

Agreed on not overloading the semantics of supported and presence.

**Reply by francescoloconte** [reply]
Your suggestions are well taken and I have applied them to the draft above.

**Comment by francescoloconte** [comment]
Rephrased section 7 to clarify that any attribute may be overridden with exceptions.

**Comment by francescoloconte** [comment]
Does anyone think that Scenario names should be declared before being used? I.e. should we add a section in Orchestra called "Scenarios", just like we have Sections and Categories? Some of the advantages would be:

1. Ease of use and check for consistency: if a scenario is to be declared in this section, when element scenarios are created they are guaranteed to use one of the declared names, rather than any name (or misspelled names). Further, since scenario names would be declared, one could check for consistency to make sure they are correct.
2. One would be able to easily see the list of all existing scenario (names). Without it, the only way to know what scenarios exists in an Orchestra file is to inspect the whole file. For example, if a team goes about creating many scenarios of elements (Equities, MiFID, FX, FO, etc.), there is no way to easily see this list. One has to traverse the whole (potentially big) file and collect them.
3. It may be possible in future to assign attributes to scenarios by setting the attributes of each scenario in this section. For example, whether a scenario (i.e. all elements belonging to that scenario) are related to a business Category or Section, or whether they are internal to the firm and not to be shared externally, etc.

Thoughts?

**Reply by JoanPuig** [reply]
I agree, that would be very useful.

**Reply by kleihan** [reply]
We already added a section called `<scenarios>` to the schema of Orchestra V1.1 RC1. However, the text in Chapter 3.5 should be improved as it is quite terse. It currently only says:
> Each scenario is described by a `<scenario>` element, a child of `<scenarios>`.

The schema for v1.1 RC1 has the following first degree elements, it includes scenarios:
```xml
				<xs:element name="metadata" type="dcterms:elementOrRefinementContainer"/>
				<xs:element ref="fixr:categories" minOccurs="0"/>
				<xs:element ref="fixr:sections" minOccurs="0"/>
				<xs:element ref="fixr:datatypes"/>
				<xs:element ref="fixr:codeSets" minOccurs="0"/>
				<xs:element ref="fixr:fields"/>
				<xs:element ref="fixr:actors" minOccurs="0"/>
				<xs:element ref="fixr:components" minOccurs="0"/>
				<xs:element ref="fixr:groups" minOccurs="0"/>
				<xs:element ref="fixr:messages"/>
				<xs:element ref="fixr:concepts" minOccurs="0"/>
				<xs:element ref="fixr:scenarios" minOccurs="0"/>
				<xs:element name="annotation" type="fixr:annotation" minOccurs="0"/>

```
Each scenario currently has an ID , a name and can have one or more annotations:
```xml
	<xs:complexType name="scenarioType">
		<xs:annotation>
			<xs:documentation>The use case of an element, distinguished by workflow, asset class, etc.</xs:documentation>
		</xs:annotation>
		<xs:sequence>
			<xs:element name="annotation" type="fixr:annotation" minOccurs="0"/>
		</xs:sequence>
		<xs:attribute name="id" type="fixr:id_t" default="1">
			<xs:annotation>
				<xs:documentation>Unique numeric identifier. Default is '1' is for base scenario.</xs:documentation>
			</xs:annotation>
		</xs:attribute>
		<xs:attribute name="name" type="fixr:Name_t" default="base">
			<xs:annotation>
				<xs:documentation>Unique name</xs:documentation>
			</xs:annotation>
		</xs:attribute>
	</xs:complexType>
```
I agree that all scenarios used for elements should be listed in the section `<scenarios>` for a valid Orchestra XML file.

**Comment by francescoloconte** [comment]
Changed "List of Abbreviations" to "List of Terms".

**Reply by francescoloconte** [reply]
Fixed an error:

`<fixr:field name="Symbol" id="55" implLength="4" scenarioId="2" inheritsScenario="base">`

We don’t support referencing the parent scenario by name.

`<fixr:field name="Symbol" id="55" implLength="4" scenarioId="2" inheritsScenarioId="1">`

**Comment by francescoloconte** [comment]
I fixed an error as below:

![Screenshot_314](https://github.com/user-attachments/assets/127aa368-9e4d-40fb-91d8-29a2d71b77cf)
